README.md
**[API docs](http://ptaoussanis.github.io/carmine/)** | **[CHANGELOG](https://github.com/ptaoussanis/carmine/blob/master/CHANGELOG.md)** | [contact & contributing](#contact--contributing) | [other Clojure libs](https://www.taoensso.com/clojure-libraries) | [Twitter](https://twitter.com/#!/ptaoussanis) | current [semantic](http://semver.org/) version:

```clojure
[com.taoensso/carmine "2.3.1"] ; Stable; see CHANGELOG for changes since 1.x
[com.taoensso/carmine "2.4.0-beta2"] ; Development; see CHANGELOG for details
```

v2 adds API improvements, integration with [Nippy v2](https://github.com/ptaoussanis/nippy) for pluggable compression+crypto, improved performance, additional message queue features, and [Tundra](#tundra) - an API for archiving cold data to an additional datastore. (A [Faraday DynamoDB](https://github.com/ptaoussanis/faraday) implementation is included).

This is a **mostly** backwards-compatible release. See the [CHANGELOG](https://github.com/ptaoussanis/carmine/blob/master/CHANGELOG.md) for migration details.

# Carmine, a Clojure Redis client & message queue

[Redis](http://www.redis.io/) is _awesome_ and it's getting [more awesome](http://www.redis.io/commands/eval) [every day](http://redis.io/topics/cluster-spec). It deserves a great Clojure client.

##### Aren't there already a bunch of clients?

Plenty: there's [redis-clojure](https://github.com/tavisrudd/redis-clojure), [clj-redis](https://github.com/mmcgrana/clj-redis) based on [Jedis](https://github.com/xetorthio/jedis), [Accession](https://github.com/abedra/accession), and (the newest) [labs-redis-clojure](https://github.com/wallrat/labs-redis-clojure). Each has its strengths but these strengths often fail to overlap, leaving one with no easy answer to an obvious question: _which one should you use_?

Carmine is an attempt to **cohesively bring together the best bits from each client**. And by bringing together the work of others I'm hoping to encourage more folks to **pool their efforts** and get behind one banner. (Rah-rah and all that).

## What's in the box™?
  * Small, uncomplicated **all-Clojure** library.
  * **Fully documented**, up-to-date API, with full support for the latest Redis versions.
  * **Great performance**.
  * Industrial strength **connection pooling**.
  * Composable, **first-class command functions**.
  * Flexible, high-performance **binary-safe serialization** using [Nippy](https://github.com/ptaoussanis/nippy).
  * Full support for **Lua scripting**, **Pub/Sub**, etc.
  * Full support for custom **reply parsing**.
  * **Command helpers** (`atomic`, `lua`, `sort*`, etc.).
  * **Ring session-store**.
  * Simple, high-performance **message queue** (Redis 2.6+, stable v2+).
  * Simple, high-performance **distributed lock** (Redis 2.6+, stable v2+).
  * Pluggable **compression** and **encryption** support. (v2+)
  * Includes _Tundra_, an API for **archiving cold data to an additional datastore**. (Redis 2.6+, v2+)

## Getting started

### Dependencies

Add the necessary dependency to your [Leiningen](http://leiningen.org/) `project.clj` and `require` the library in your ns:

```clojure
[com.taoensso/carmine "2.3.1"] ; project.clj
(ns my-app (:require [taoensso.carmine :as car :refer (wcar)])) ; ns
```

### Connections

You'll usually want to define a single connection pool, and one connection spec for each of your Redis servers.

```clojure
(def server1-conn {:pool {<opts>} :spec {<opts>}})
(defmacro wcar* [& body] `(car/wcar server1-conn ~@body))
```

###### See the relevant docstrings for pool+spec options

### Basic commands

Sending commands is easy:

```clojure
(wcar* (car/ping)
       (car/set "foo" "bar")
       (car/get "foo"))
=> ["PONG" "OK" "bar"]
```

Note that sending multiple commands at once like this will employ [pipelining](http://redis.io/topics/pipelining). The replies will be queued server-side and returned all at once as a vector.

If the server responds with an error, an exception is thrown:

```clojure
(wcar* (car/spop "foo"))
=> Exception ERR Operation against a key holding the wrong kind of value
```

But what if we're pipelining?

```clojure
(wcar* (car/set  "foo" "bar")
       (car/spop "foo")
       (car/get  "foo"))
=> ["OK" #<Exception ERR Operation against ...> "bar"]
```

### Serialization

The only value type known to Redis internally is the [byte string](http://redis.io/topics/data-types). But Carmine uses [Nippy](https://github.com/ptaoussanis/nippy) under the hood and understands all of Clojure's [rich datatypes](http://clojure.org/datatypes), letting you use them with Redis painlessly:

```clojure
(wcar* (car/set "clj-key" {:bigint (bigint 31415926535897932384626433832795)
                           :vec    (vec (range 5))
                           :set    #{true false :a :b :c :d}
                           :bytes  (byte-array 5)
                           ;; ...
                           })
       (car/get "clj-key"))
=> ["OK" {:bigint 31415926535897932384626433832795N
          :vec    [0 1 2 3 4]
          :set    #{true false :a :c :b :d}
          :bytes  #<byte [] [B@4d66ea88>}]
```

Types are handled as follows:
 * Clojure strings become Redis strings.
 * Keywords become Redis strings. (v2+)
 * Simple Clojure numbers (integers, longs, floats, doubles) become Redis strings.
 * Everything else gets automatically de/serialized.

You can force automatic de/serialization for an argument of any type by wrapping it with `car/serialize`.

### Documentation and command coverage

Like [labs-redis-clojure](https://github.com/wallrat/labs-redis-clojure), Carmine uses the [official Redis command reference](https://github.com/antirez/redis-doc/blob/master/commands.json) to generate its own command API. Which means that not only is Carmine's command coverage *always complete*, but it's also **fully documented**:

```clojure
(use 'clojure.repl)
(doc car/sort)
=> "SORT key [BY pattern] [LIMIT offset count] [GET pattern [GET pattern ...]] [ASC|DESC] [ALPHA] [STORE destination]

Sort the elements in a list, set or sorted set.

Available since: 1.0.0.

Time complexity: O(N+M*log(M)) where N is the number of elements in the list or set to sort, and M the number of returned elements. When the elements are not sorted, complexity is currently O(N) as there is a copy step that will be avoided in next releases."
```

*Yeah*. Andreas Bielk, you rock.

### Lua

Redis 2.6 introduced a remarkably powerful feature: server-side Lua scripting! As an example, let's write our own version of the `set` command:

```clojure
(defn my-set
  [key value]
  (car/lua "return redis.call('set', _:my-key, 'lua '.. _:my-val)"
                  {:my-key key}   ; Named key variables and their values
                  {:my-val value} ; Named non-key variables and their values
                  ))

(wcar* (my-set "foo" "bar")
       (car/get "foo"))
=> ["OK" "lua bar"]
```

Script primitives are also provided: `eval`, `eval-sha`, `eval*`, `eval-sha*`. See the [Lua scripting docs](http://redis.io/commands/eval) for more info.

### Helpers

The `lua` command above is a good example of a Carmine *helper*.

Carmine will never surprise you by interfering with the standard Redis command API. But there are times when it might want to offer you a helping hand (if you want it). Compare:

```clojure
(wcar* (car/zunionstore "dest-key" 3 "zset1" "zset2" "zset3" "WEIGHTS" 2 3 5))
(wcar* (car/zunionstore* "dest-key" ["zset1" "zset2" "zset3"] "WEIGHTS" 2 3 5))
```

Both of these calls are equivalent but the latter counted the keys for us. `zunionstore*` is another helper: a slightly more convenient version of a standard command, suffixed with a `*` to indicate that it's non-standard.

Helpers currently include: `atomic`, `eval*`, `evalsha*`, `hgetall*`, `info*`, `lua`, `sort*`, `zinterstore*`, and `zunionstore*`. See their docstrings for more info.

### Commands are (just) functions

In Carmine, Redis commands are *real functions*. Which means you can *use* them like real functions:

```clojure
(wcar* (doall (repeatedly 5 car/ping)))
=> ["PONG" "PONG" "PONG" "PONG" "PONG"]

(let [first-names ["Salvatore"  "Rich"]
      surnames    ["Sanfilippo" "Hickey"]]
  (wcar* (mapv #(car/set %1 %2) first-names surnames)
         (mapv car/get first-names)))
=> ["OK" "OK" "Sanfilippo" "Hickey"]

(wcar* (mapv #(car/set (str "key-" %) (rand-int 10)) (range 3))
       (mapv #(car/get (str "key-" %)) (range 3)))
=> ["OK" "OK" "OK" "OK" "0" "6" "6" "2"]
```

And since real functions can compose, so can Carmine's. By nesting `wcar` calls, you can fully control how composition and pipelining interact:

```clojure
(let [hash-key "awesome-people"]
  (wcar* (car/hmset hash-key "Rich" "Hickey" "Salvatore" "Sanfilippo")
         (mapv (partial car/hget hash-key)
               ;; Execute with own connection & pipeline then return result
               ;; for composition:
               (wcar* (car/hkeys hash-key)))))
=> ["OK" "Sanfilippo" "Hickey"]
```

### Listeners & Pub/Sub

Carmine has a flexible **Listener** API to support persistent-connection features like [monitoring](http://redis.io/commands/monitor) and Redis's fantastic [Publish/Subscribe](http://redis.io/topics/pubsub) feature:

```clojure
(def listener
  (car/with-new-pubsub-listener (:spec server1-conn)
    {"foobar" (fn f1 [msg] (println "Channel match: " msg))
     "foo*"   (fn f2 [msg] (println "Pattern match: " msg))}
   (car/subscribe  "foobar" "foobaz")
   (car/psubscribe "foo*")))
```

Note the map of message handlers. `f1` will trigger when a message is published to channel `foobar`. `f2` will trigger when a message is published to `foobar`, `foobaz`, `foo Abraham Lincoln`, etc.

Publish messages:

```clojure
(wcar* (car/publish "foobar" "Hello to foobar!"))
```

Which will trigger:

```clojure
(f1 '("message" "foobar" "Hello to foobar!"))
;; AND ALSO
(f2 '("pmessage" "foo*" "foobar" "Hello to foobar!"))
```

You can adjust subscriptions and/or handlers:

```clojure
(with-open-listener listener
  (car/unsubscribe) ; Unsubscribe from every channel (leave patterns alone)
  (car/psubscribe "an-extra-channel"))

(swap! (:state listener) assoc "*extra*" (fn [x] (println "EXTRA: " x)))
```

**Remember to close the listener** when you're done with it:

```clojure
(car/close-listener listener)
```

Note that subscriptions are *connection-local*: you can have three different listeners each listening for different messages, using different handlers. This is great stuff.

### Reply parsing


CHANGELOG.md
## v2.3.1 → v2.4.0-beta2 (major update!)

  * **IMPORTANT** Message queues: pre-2.4.0-beta1 queues *should* be compatible with 2.4.0-beta1+, but I would recommend **draining your old queues before upgrading** to 2.4.0-beta1+ to be on the safe side. That is: if you have any queued work outstanding - finish processing the work **before upgrading Carmine**.
  * **POTENTIALLY BREAKING** Parsers: `with-replies` now passes enclosing parser state to its body. This only affects you if you have used `with-replies` to write custom Redis commands that operate within an implicit context _and_ you interpret the `with-replies` result internally. The new docstring contains details.

  * Parsers: completely refactored design for robustness+flexibility.
  * Parsers: new unit-test suite.
  * Parsers: fixed a number of subtle bugs, mostly internal.
  * Parsers: Added `parser-comp` fn for composing parsers (see docstring for details).

  * Message queues: completely refactored design for robustness+efficiency.
  * Message queues: new unit-test suite.
  * Message queues: fixed a number of bugs, mostly garbage-collection issues.
  * Message queues: `enqueue` now accepts an optional, custom unique message id (e.g. message hash).
  * Message queues: handlers may now return {:status :success :backoff-ms `<msecs>`} for de-duplicate backoff. `enqueue` will return an error in these cases (see docstring for details).
  * Message queues: `message-status` can now be called within pipelines.
  * Message queues: `message-status` now has more detailed return types (see docstring for details).
  * Message queues: multi-worker end-of-queue backoffs are now synchronized more efficiently.
  * Message queues: workers now accept an optional monitor fn for queue-status logging, etc. (see docstring for details). A default monitor is provided that will warn when queue size > 1000 items.
  * Message queues: workers now accept an optional `nthreads` arg for spinning up synchronized multi-threaded workers.
  * Message queues: handlers now receive message id along with other args.
  * Message queues: `enqueue` now takes an optional flag to specify dedupe behaviour when message is locked.


## v2.2.3 → v2.3.1
  * **DEPRECATED**: `atomically`, `ensure-atomically` -> `atomic`. The new macro is faster, more robust, more flexible. See docstring for details.
  * Official Redis command fns now contain a `:redis-api` metadata key that describes the first version of Redis to support the command.


## v2.2.0 → v2.2.3
  * Fix race condition for pool creation (thanks cespare!).
  * Fix unnecessary reflection for pool creation (thanks harob!).


## v2.1.4 → v2.2.0
  * Add `hmset*`, `hmget*` helpers.
  * Add `:clojurize?` option to `info*` helper.
  * Allow `hmget*`, `hgetall*`, `zinterstore*` to work with custom parsers.


## v2.1.0 → v2.1.4
  * Fixed `lua` clashing var name regex bug (thanks to Alex Kehayias for report).


## v2.0.0 → v2.1.0
  * Like `with-replies`, `wcar` macro can now take a first `:as-pipeline` arg:
  ```clojure
  (wcar {} (car/ping)) => "PONG"
  (wcar {} :as-pipeline (car/ping)) => ["PONG"]
  ```
  * **DEPRECATED**: `kname` -> `key`. The new fn does NOT automatically filter input parts for nil. This plays better with Redis' key pattern matching style, but may require manual filtering in some cases.


## v1.12.0 → v2.0.0

  * Refactored a bunch of code for simplicity+performance (~20% improved roundtrip times).
  * Upgraded to [Nippy v2][Nippy GitHub] for pluggable compression+crypto. See the [Nippy CHANGELOG][] for details.
  * Added early (alpha) Tundra API for semi-automatic cold data archiving. See the [README](https://github.com/ptaoussanis/carmine#tundra) for details.

  * **DEPRECATED**: `with-conn`, `make-conn-pool`, `make-conn-spec` -> `wcar`:
  ```clojure
  ;;; Old usage
  (def conn-pool (car/make-conn-pool <opts>))
  (def conn-spec (car/make-conn-spec <opts>))
  (defmacro wcar* [& body] `(car/with-conn conn-pool conn-spec ~@body))

  ;;; New idiomatic usage
  (ns my-app (:require [taoensso.carmine :as car :refer (wcar)]))
  (def server1-conn {:pool {<opts>} :spec {<opts>}})
  (wcar server1-conn (car/ping)) => "PONG"

  ;; or
  (defmacro wcar* [& body] `(car/wcar server1-conn ~@body))
  ```

  * **BREAKING**: Keyword args now get stringified instead of serialized:
    ```clojure
    (wcar* (car/set "foo" :bar) (car/get "foo")) => :bar  ; Old behaviour
    (wcar* (car/set "foo" :bar) (car/get "foo")) => "bar" ; New

    ;; Idiomatic usage is now:
    (wcar* (car/set "foo" :bar) (car/parse-keyword (car/get "foo"))) => :bar

    ;; Or use `freeze` if you want to serialize the input arg like before:
    (wcar* (car/set "foo" (car/freeze :bar)) (car/get "foo")) => :bar
    ```

  * **BREAKING**: Raw binary data return type is now unwrapped:
    ```clojure
    (wcar* (car/get "raw-bytes")) => [<length> [B@4d66ea88>]] ; Old behaviour
    (wcar* (car/get "raw-bytes")) => [B@4d66ea88> ; New
    ```

  * **BREAKING**: Distributed locks API has changed:
    The old API used an atom for connection config. The new API takes an explicit `conn` arg for every fn/macro.

  * Added message queue per-message backoffs:
    ```clojure
    ;; Handler fns can now return a response of form:
    {:status     <#{:success :error :retry}>

src/taoensso/carmine/message_queue.clj
(ns taoensso.carmine.message-queue
  "Carmine-backed Clojure message queue. All heavy lifting by Redis (2.6+).
  Simple implementation. Very simple API. Reliable. Fast.

  Redis keys:
    * carmine:mq:<qname>:messages     -> hash, {mid mcontent}.
    * carmine:mq:<qname>:locks        -> hash, {mid lock-expiry-time}.
    * carmine:mq:<qname>:backoffs     -> hash, {mid backoff-expiry-time}.
    * carmine:mq:<qname>:nattempts    -> hash, {mid attempt-count}.
    * carmine:mq:<qname>:mid-circle   -> list, rotating list of mids.
    * carmine:mq:<qname>:gc           -> set, for efficient mid removal from circle.
    * carmine:mq:<qname>:requeue      -> set, for `allow-locked-dupe?` enqueue.
    * carmine:mq:<qname>:eoq-backoff? -> ttl flag, used for queue-wide (every-worker)
                                         polling backoff.
    * carmine:mq:<qname>:ndry-runs    -> int, number of times worker(s) have burnt
                                         through queue w/o work to do.

  Ref. http://antirez.com/post/250 for basic implementation details."
  {:author "Peter Taoussanis"}
  (:require [clojure.string   :as str]
            [taoensso.carmine :as car :refer (wcar)]
            [taoensso.timbre  :as timbre]))

;; TODO Redis 2.8+ Pub/Sub notifications could be used to more efficiently
;; coordinate worker backoffs. This'd allow us to essentially go from a polling
;; design to an evented design even w/o the need for a Lua brpoplpush.
;;
;; Note that we'd want the Pub/Sub notifier _in addition to_ the current
;; rpoplpush design: both for atomicity and reliability (Pub/Sub doesn't
;; currently have reliability guarantees).

;;;; Public utils

(defn exp-backoff "Returns binary exponential backoff value."
  [attempt & [{:keys [factor min max] :or {factor 2200}}]]
  (let [binary-exp (Math/pow 2 (dec attempt))
        time (* (+ binary-exp (rand binary-exp)) 0.5 factor)]
    (long (let [time (if min (clojure.core/max min time) time)
                time (if max (clojure.core/min max time) time)]
            time))))

(comment (mapv #(exp-backoff % {}) (range 5)))

;;;; Admin

(def qkey "Prefixed queue key" (memoize (partial car/key :carmine :mq)))

(defn clear-queues [conn & qnames]
  (wcar conn
    (doseq [qname qnames]
      (when-let [qks (seq (wcar conn (car/keys (qkey qname :*))))]
        (apply car/del qks)))))

(defn queue-status [conn qname]
  (let [qk (partial qkey qname)]
    (zipmap [:messages :locks :backoffs :nattempts :mid-circle :gc
             :eoq-backoff? :ndry-runs :requeue]
     (wcar conn
       (car/hgetall*      (qk :messages))
       (car/hgetall*      (qk :locks))
       (car/hgetall*      (qk :backoffs))
       (car/hgetall*      (qk :nattempts))
       (car/lrange        (qk :mid-circle) 0 -1)
       (->> (car/smembers (qk :gc))           (car/parse set))
       (->> (car/smembers (qk :requeue))      (car/parse set))
       (->> (car/get      (qk :eoq-backoff?)) (car/parse-bool)) ; Give TTL?
       (->> (car/get      (qk :ndry-runs))    (car/parse-long))))))

;;;; Implementation

(defn message-status
  "Returns current message status, e/o:
    :queued            - Awaiting (re)handling.
    :locked            - Currently with handler.
    :retry-backoff     - Awaiting backoff for handler retry (rehandling).
    :done-with-backoff - Finished handling, awaiting dedupe timeout.
    :done-awaiting-gc  - Finished handling, no dedupe timeout, awaiting GC.
    nil                - Already GC'd or invalid message id."
  [qname mid]
  (car/parse-keyword
   (car/lua ; Careful, logic here is necessarily quite subtle!
    "--
    local now         = tonumber(_:now)
    local lock_exp    = tonumber(redis.call('hget', _:qk-locks,    _:mid) or 0)
    local backoff_exp = tonumber(redis.call('hget', _:qk-backoffs, _:mid) or 0)

    --- Note: `else`s necessary for `enqueue`
    if redis.call('hexists', _:qk-messages, _:mid) ~= 1 then
      if (now < backoff_exp) then return 'done-with-backoff' end
      return nil
    else
      if redis.call('sismember', _:qk-gc, _:mid) == 1 then
        if (now < backoff_exp) then return 'done-with-backoff' end
        return 'done-awaiting-gc'
      else
        if     (now < lock_exp)    then return 'locked'
        elseif (now < backoff_exp) then return 'retry-backoff'
        else
          return 'queued'
        end
      end
    end"
    {:qk-messages (qkey qname :messages)
     :qk-locks    (qkey qname :locks)
     :qk-backoffs (qkey qname :backoffs)
     :qk-gc       (qkey qname :gc)}
    {:now (System/currentTimeMillis)
     :mid mid})))

(defn enqueue
  "Pushes given message (any Clojure datatype) to named queue and returns unique
  message id or {:carmine.mq/error #{<:queued :locked :retry-backoff
                                      :done-with-backoff>}}.
  Options:
    unique-message-id  - Specify an explicit message id (e.g. message hash) to
                         perform a de-duplication check. If unspecified, a unique
                         message id will be auto-generated.
    allow-locked-dupe? - Alpha - subject to change.
                         When true, a locked message will NOT prevent an
                         additional message with the same id from being queued.
                         Actually, this message will be held in escrow and
                         queued (once) iff the locked message completes handling
                         successfully. This mechanism provides a tool for
                         defeating race conditions when writing handlers which
                         read Redis state that may change before handling has
                         been acknowledged by queue."
  [qname message & [unique-message-id allow-locked-dupe?]]
  (->>
   (car/lua
    "--- Check for dupes (modified `message-status` logic)
    local now         = tonumber(_:now)
    local lock_exp    = tonumber(redis.call('hget', _:qk-locks,    _:mid) or 0)
    local backoff_exp = tonumber(redis.call('hget', _:qk-backoffs, _:mid) or 0)

    if redis.call('hexists', _:qk-messages, _:mid) ~= 1 then
      if (now < backoff_exp) then return 'done-with-backoff' end
      -- return nil
    else
      if redis.call('sismember', _:qk-gc, _:mid) == 1 then
        if (now < backoff_exp) then return 'done-with-backoff' end
        -- return 'done-awaiting-gc'
      else
        if (now < lock_exp) then
          if _:allow-locked-dupe? ~= 'true' then return 'locked' end
        elseif (now < backoff_exp) then return 'retry-backoff'
        else
          return 'queued'
        end
      end
    end
    ---

    if (now < lock_exp) then
      redis.call('sadd', _:qk-requeue, _:mid)
    else
      redis.call('hset', _:qk-messages, _:mid, _:mcontent)

      -- lpushnx end-of-circle marker to ensure an initialized mid-circle
      if redis.call('exists', _:qk-mid-circle) ~= 1 then
        redis.call('lpush', _:qk-mid-circle, 'end-of-circle')
      end

      redis.call('lpush', _:qk-mid-circle, _:mid)
    end

    return {_:mid} -- Wrap to distinguish from error replies
" ; Necessary for Lua script-end weirdness
    {:qk-messages   (qkey qname :messages)
     :qk-locks      (qkey qname :locks)
     :qk-backoffs   (qkey qname :backoffs)
     :qk-mid-circle (qkey qname :mid-circle)
     :qk-gc         (qkey qname :gc)
     :qk-requeue    (qkey qname :requeue)}
    {:now           (System/currentTimeMillis)
     :mid           (or unique-message-id (str (java.util.UUID/randomUUID)))
     :mcontent      (car/freeze message)
     :allow-locked-dupe? (if allow-locked-dupe? "true" "false")})
   (car/parse #(if (vector? %) (first %) {:carmine.mq/error (keyword %)}))))

(defn dequeue
  "IMPLEMENTATION DETAIL: Use `worker` instead.
  Rotates queue's mid-circle and processes next mid. Returns:
    nil             - If msg already GC'd, locked, or set to backoff.
    \"eoq-backoff\" - If circle uninitialized or end-of-circle marker reached.
    [<mid> <mcontent> <attempt-count>] - If message should be (re)handled now."
  [qname & [{:keys [lock-ms eoq-backoff-ms]
             :or   {lock-ms (* 1000 60 60) eoq-backoff-ms exp-backoff}}]]
  (let [;; Precomp 5 backoffs so that `dequeue` can init the backoff
        ;; atomically. This is hackish, but a decent tradeoff.
        eoq-backoff-ms-vec
        (cond (fn?      eoq-backoff-ms) (mapv eoq-backoff-ms              (range 5))
              (integer? eoq-backoff-ms) (mapv (constantly eoq-backoff-ms) (range 5))
              :else (throw (Exception. (str "Bad eoq-backoff-ms: " eoq-backoff-ms))))]
    (car/lua
     "if redis.call('exists', _:qk-eoq-backoff) == 1 then return 'eoq-backoff' end

     -- TODO Waiting for Lua brpoplpush support to get us long polling
     local mid = redis.call('rpoplpush', _:qk-mid-circle, _:qk-mid-circle)
     local now = tonumber(_:now)

     if (not mid) or (mid == 'end-of-circle') then -- Uninit'd or eoq

       -- Calculate eoq_backoff_ms
       local ndry_runs = tonumber(redis.call('get', _:qk-ndry-runs) or 0)
       local eoq_ms_tab = {_:eoq-ms0, _:eoq-ms1, _:eoq-ms2, _:eoq-ms3, _:eoq-ms4}
       local eoq_backoff_ms = tonumber(eoq_ms_tab[math.min(5, (ndry_runs + 1))])

       -- Set queue-wide polling backoff flag
       redis.call('psetex', _:qk-eoq-backoff, eoq_backoff_ms, 'true')
       redis.call('incr',   _:qk-ndry-runs)

       -- Prune expired backoffs. We have to do this manually if we want to keep
       -- backoffs in a hash as we do for efficiency & Cluster support. This hash
       -- is usu. quite small since it consists only of active retry and dedupe
       -- backoffs.
       for i,k in pairs(redis.call('hkeys', _:qk-backoffs)) do
         if (now >= tonumber(redis.call('hget', _:qk-backoffs, k))) then
           redis.call('hdel', _:qk-backoffs, k)
         end
       end

       return 'eoq-backoff'
     end

     if redis.call('sismember', _:qk-gc, mid) == 1 then -- GC
       redis.call('lrem', _:qk-mid-circle, 1, mid) -- Efficient here
       redis.call('srem', _:qk-gc,            mid)
       redis.call('hdel', _:qk-messages,      mid)
       redis.call('hdel', _:qk-locks,         mid)
       --redis.call('hdel', _:qk-backoffs,    mid) -- No! Will use as dedupe backoff
       redis.call('hdel', _:qk-nattempts,     mid)
       redis.call('set',  _:qk-ndry-runs, 0) -- Did work
       return nil
     end

     local lock_exp    = tonumber(redis.call('hget', _:qk-locks,    mid) or 0)
     local backoff_exp = tonumber(redis.call('hget', _:qk-backoffs, mid) or 0)

     -- Active lock/backoff
     if (now < lock_exp) or (now < backoff_exp) then return nil end

     redis.call('hset', _:qk-locks,    mid, now + tonumber(_:lock-ms)) -- Acquire
     redis.call('hdel', _:qk-backoffs, mid) -- Expired, so prune (keep gc-prune fast)
     redis.call('set',  _:qk-ndry-runs, 0) -- Did work

     local mcontent  = redis.call('hget',    _:qk-messages,  mid)
     local nattempts = redis.call('hincrby', _:qk-nattempts, mid, 1)
     return {mid, mcontent, nattempts}"
     {:qk-messages    (qkey qname :messages)
      :qk-locks       (qkey qname :locks)
      :qk-backoffs    (qkey qname :backoffs)
      :qk-nattempts   (qkey qname :nattempts)
      :qk-mid-circle  (qkey qname :mid-circle)
      :qk-gc          (qkey qname :gc)
      :qk-eoq-backoff (qkey qname :eoq-backoff?)
      :qk-ndry-runs   (qkey qname :ndry-runs)}
     {:now            (System/currentTimeMillis)
      :lock-ms        lock-ms
      :eoq-ms0        (nth eoq-backoff-ms-vec 0)
      :eoq-ms1        (nth eoq-backoff-ms-vec 1)
      :eoq-ms2        (nth eoq-backoff-ms-vec 2)
      :eoq-ms3        (nth eoq-backoff-ms-vec 3)
      :eoq-ms4        (nth eoq-backoff-ms-vec 4)})))

(comment
  (clear-queues {} :q1)
  (queue-status {} :q1)
  (wcar {} (enqueue :q1 :msg1 :mid1))
  (wcar {} (message-status :q1 :mid1))
  (wcar {} (dequeue :q1))
  ;;(mapv exp-backoff (range 5))
  (wcar {} (car/pttl (qkey :q1 :eoq-backoff?))))

(defn handle1 "Implementation detail!"
  [conn qname handler [mid mcontent attempt :as poll-reply]]
  (when (and poll-reply (not= poll-reply "eoq-backoff"))
    (let [qk   (partial qkey qname)
          done (fn [status mid & [backoff-ms]]
                 (car/atomic conn 100
                   (car/watch (qk :requeue))
                   (let [requeue?
                         (wcar {} (->> (car/sismember (qk :requeue) mid)
                                       (car/parse-bool)))
                         status (if (and (= status :success) requeue?)
                                  :requeue status)]
                     (car/multi)
                     (when backoff-ms ; Retry or dedupe backoff, depending on type
                       (car/hset (qk :backoffs) mid (+ (System/currentTimeMillis)
                                                       backoff-ms)))
                     (case status
                       (:success :error) (car/sadd (qk :gc) mid) ; Queue for GC
                       (:retry :requeue)
                       (do (car/hdel (qk :locks)   mid) ; Unlock
                           (car/srem (qk :requeue) mid)
                           (when (= status :requeue)
                             (car/hdel (qk :nattempts) mid)))))))

          error (fn [mid poll-reply & [throwable]]
                  (done mid :error)
                  (timbre/errorf
                   (if throwable throwable (Exception. ":error handler response"))
                   "Error handling %s queue message:\n%s" qname poll-reply))

          {:keys [status throwable backoff-ms]}
          (let [result (try (handler {:mid mid :message mcontent :attempt attempt})
                            (catch Throwable t {:status :error :throwable t}))]
            (when (map? result) result))]

      (case status
        :success (done status mid backoff-ms)
        :retry   (done status mid backoff-ms)
        :error   (error mid poll-reply throwable)
        (do (done :success mid) ; For backwards-comp with old API
            (timbre/warnf "Invalid handler status: %s" status))))))

;;;; Workers

(defprotocol IWorker (start [this]) (stop [this]))
(defrecord    Worker [conn qname running? opts]
  java.io.Closeable (close [this] (stop this))
  IWorker
  (stop  [_] (let [stopped? @running?] (reset! running? false) stopped?))
  (start [_]
    (when-not @running?
      (reset! running? true)
      (let [{:keys [handler monitor nthreads throttle-ms]} opts
            qk (partial qkey qname)
            start-polling-loop!
            (fn []
              (try
                (while @running?
                  (let [[poll-reply ndruns mid-circle-size]
                        (wcar conn (dequeue qname opts)
                                   (car/get  (qk :ndry-runs))
                                   (car/llen (qk :mid-circle)))]

                    (when monitor (monitor {:mid-circle-size mid-circle-size
                                            :ndry-runs       (or ndruns 0)
                                            :poll-reply      poll-reply}))

                    (if (= poll-reply "eoq-backoff")
                      (Thread/sleep (max (wcar conn (car/pttl (qk :eoq-backoff?)))
                                         10))
                      (handle1 conn qname handler poll-reply)))
                  (when throttle-ms (Thread/sleep throttle-ms)))
                (catch Throwable t (timbre/fatalf t "Worker error!") (throw t))))]
        (dorun (repeatedly nthreads (fn [] (future (start-polling-loop!))))))
      true)))

(defn monitor-fn
  "Returns a worker monitor fn that warns when queue's mid-circle exceeds
  the prescribed size. A backoff timeout can be provided to rate-limit this
  warning."
  [qname max-circle-size warn-backoff-ms]
  (let [udt-last-warning (atom nil)]
    (fn [{:keys [mid-circle-size]}]
      (let [instant (System/currentTimeMillis)]
        (when (and (> mid-circle-size max-circle-size)
                   (> (- instant (or @udt-last-warning 0))
                      (or warn-backoff-ms 0)))
          (reset! udt-last-warning instant)
          (timbre/warnf "Message queue size warning: %s (mid-circle-size: %s)"
                        qname max-circle-size))))))

(defn worker
  "Returns a threaded worker to poll for and handle messages `enqueue`'d to
  named queue. Options:
   :handler        - (fn [{:keys [mid message attempt]}]) that throws an exception
                     or returns {:status     <#{:success :error :retry}>
                                 :throwable  <Throwable>
                                 :backoff-ms <retry-or-dedupe-backoff-ms}.
   :monitor        - (fn [{:keys [mid-circle-size ndry-runs poll-reply]}]) called
                     on each worker loop iteration. Useful for queue
                     monitoring/logging. See also `monitor-fn`.
   :lock-ms        - Max time handler may keep a message before handler
                     considered fatally stalled and message re-queued. Must be
                     sufficiently high to prevent double handling.
   :eoq-backoff-ms - Thread sleep period each time end of queue is reached.
                     Can be a (fn [ndry-runs]) -> ms (n<=5) will be used.
                     Sleep synchronized for all queue workers.
   :nthreads       - Number of synchronized worker threads to use.
   :throttle-ms    - Thread sleep period between each poll."
  [conn qname & [{:keys [handler monitor lock-ms eoq-backoff-ms nthreads
                         throttle-ms auto-start?]
                  :or   {handler (fn [args] (timbre/infof "%s" args)
                                           {:status :success})
                         monitor (monitor-fn qname 1000 (* 1000 60 60 6))
                         lock-ms        (* 1000 60 60)
                         nthreads       1
                         throttle-ms    200
                         eoq-backoff-ms exp-backoff
                         auto-start?    true}}]]
  (let [w (->Worker conn qname (atom false)
                    {:handler        handler
                     :monitor        monitor
                     :lock-ms        lock-ms
                     :eoq-backoff-ms eoq-backoff-ms
                     :nthreads       nthreads
                     :throttle-ms    throttle-ms})]
    (when auto-start? (start w)) w))

;;;; Renamed/deprecated

(defn make-dequeue-worker "DEPRECATED: Use `worker` instead."
  [pool spec & {:keys [handler-fn handler-ttl-msecs backoff-msecs throttle-msecs
                       auto-start?]}]
  (worker {:pool pool :spec spec}
    (merge (when-let [ms handler-ttl-msecs] {:lock-ms        ms})
           (when-let [ms backoff-msecs]     {:eoq-backoff-ms ms})
           (when-let [ms throttle-msecs]    {:throttle-ms    ms})
           (when-let [hf handler-fn]
             {:handler (fn [{:keys [message]}]
                         {:status (or (#{:success :error :retry} (hf message))
                                      :success)})})
           {:auto-start? auto-start?})))
